docs/design/storage_abstraction.md

```diff
@@ -0,0 +1,275 @@
+
+    public(friend) fun contains<K: copy + drop, V>(table_handle: address, key: K): bool;
+}
+```
+
+TableHandle is the ObjectID of the TableObject. The upper-level data structure of RawTable can directly operate on any type of K, V, like a raw-level KV storage interface. Each RowTable corresponds to a SMT in the StateDB. The ObjectID of the global ObjectStore is a special ObjectID with all bytes set to `0`.
+
+Currently, all methods in RawTable are `friend` methods, and developers cannot directly operate on them. They need to operate through the upper-level data structures that encapsulate RawTable, such as `Table<K,V>`, `TypeTable`, and `ObjectStorage`.
+
+### Rooch Object
+
+Objects in Rooch adopt a Box mode where they are equivalent to a unique ID box that can encapsulate a type `T` to create an Object.
+
+```move
+module moveos_std::object{
+    struct Object<T: key>{
+        id: ObjectID,
+        owner: address,
+        value: T,
+    }
+}
+```
+
+Rooch Objects are also a use case for the [Hot Potato](https://examples.sui.io/patterns/hot-potato.html) pattern in Move. Objects do not have any `ability`, so they cannot be `drop`, `copy`, or `store`, and can only be handled by ObjectStorage API after creation.
+
+The advantage of this design is that no additional requirements need to be placed on the definition of Move's Struct, nor is additional support of Native methods required, since the concept of Object can be defined directly in Move. ObjectStorage is a storage space outside of a user's storage space, with a unique global ID as the key, to address situations in which global storage is needed.
+
+### ObjectStorage
+
+ObjectStorage provides interfaces related to global storage of Objects. It is implemented based on RawTable and is defined as follows:
+
+```move
+module moveos_std::object_storage{
+    #[private_generics(T)]
+    /// Borrow Object from object store with object_id
+    public fun borrow<T: key>(this: &ObjectStorage, object_id: ObjectID): &Object<T>;
+
+    #[private_generics(T)]
+    /// Borrow mut Object from object store with object_id
+    public fun borrow_mut<T: key>(this: &mut ObjectStorage, object_id: ObjectID): &mut Object<T>;
+
+    #[private_generics(T)]
+    /// Remove object from object store
+    public fun remove<T: key>(this: &mut ObjectStorage, object_id: ObjectID): Object<T>;
+
+    #[private_generics(T)]
+    /// Add object to object store
+    public fun add<T: key>(this: &mut ObjectStorage, obj: Object<T>);
+
+    public fun contains<T: key>(this: &mut ObjectStorage, object_id: ObjectID): bool;
+}
+```
+
+The above methods are ensured to only be called directly by Modules containing `T` through [private_generics](https://github.com/rooch-network/rooch/issues/64), and the safety model follow to the constraints of the Move global storage instructions.
+
+>TBD: Whether to provide a method that allows Owners to directly operate on Objects, similar to Sui's `public_transfer`, needs further study.
```

Comment:
I think yes cause on of the benefit of objects is that I can use a native function to fastly and inexpensively transfers it.
If I have objects and I still need to implement a rather complicated methods to transfer it, it might cost me more gas fee?

Reply:
This design focuses on the Object's expressive ability rather than execution efficiency, So make it TBD.